jsonrpc2.go

```diff
@@ -675,28 +695,7 @@ func (c *Conn) readMessages(ctx context.Context) {
 			}
 		}
 	}
-
-	c.sending.Lock()
-	c.mu.Lock()
-	c.shutdown = true
-	closing := c.closing
-	if err == io.EOF {
-		if closing {
-			err = ErrClosed
-		} else {
-			err = io.ErrUnexpectedEOF
-		}
-	}
-	for _, call := range c.pending {
-		call.done <- err
-		close(call.done)
-	}
-	c.mu.Unlock()
-	c.sending.Unlock()
-	if err != io.ErrUnexpectedEOF && !closing {
-		c.logger.Printf("jsonrpc2: protocol error: %v\n", err)
-	}
-	close(c.disconnect)
+	c.close(err)
```

Comment:
This is the source of the bug that's being fixed. Before, `close(c.disconnect)` was being called here but `c.stream.Close()` was not. `c.stream.Close()` was only being called if `Conn` was explicitly being closed by `Conn.Close()`.

I rearranged the code so that the "closing" procedure is the same regardless if the cause is an error in `Conn.readMessages` or if the cause is a call to `Conn.Close()`. This should ensure that the state of the underlying connection (i.e. `c.stream`) matches what `Conn.DiconnectNotify()` reports.

From a high level, this is what it looks like:

```go
func (c *Conn) readMessages() {
   var err error
   for err == nil {
     // read from underlying connection
   }
   c.close(err)
}

func (c *Conn) Close() error {
   return c.close(nil)
}

func (c *Conn) close(cause err) error {
   // common closing procedure
}

```